handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -1800,17 +1800,24 @@ private boolean setHandshakeSuccessIfStillHandshaking() {
      * Notify all the handshake futures about the successfully handshake
      */
     private void setHandshakeSuccess() {
-        handshakePromise.trySuccess(ctx.channel());
-
-        if (logger.isDebugEnabled()) {
-            SSLSession session = engine.getSession();
-            logger.debug(
-              "{} HANDSHAKEN: protocol:{} cipher suite:{}",
-              ctx.channel(),
-              session.getProtocol(),
-              session.getCipherSuite());
+        boolean notified = handshakePromise.trySuccess(ctx.channel());
+        SSLSession session = engine.getSession();
+
+        // There seems to be a bug in the SSLEngineImpl that is part of the OpenJDK that results in returning
+        // HandshakeStatus.FINISHED multiple times which is not expected. This only happens in TLSv1.3 so lets
+        // ensure we only notify once in this case.
+        //
+        // This is safe as TLSv1.3 does not support renegotiation and so we should never see two handshake events.
+        if (notified || engineType == SslEngineType.JDK && !SslUtils.PROTOCOL_TLS_V1_3.equals(session.getProtocol())) {
```

Comment:
Is `engineType == SslEngineType.JDK` is necessary here? I think we can have only `notified || !SslUtils.PROTOCOL_TLS_V1_3.equals(session.getProtocol())`.
Also, IIUC in case of renegotiation netty initializes a new `handshakePromise`. So, the `notified` will also be `true` for renegotiation.

Reply:
I guess yeah we can simplify. This won't harm